homeassistant/components/zha/radio_manager.py

```diff
@@ -156,3 +200,130 @@ async def async_reset_adapter(self) -> None:
+            return False
+
+        try:
+            await self._hass.config_entries.async_unload(self._config_entry.entry_id)
+        except config_entries.OperationNotAllowed:
+            # ZHA is not running
+            pass
+
+        # Load our current network settings
+        await self._radio_mgr.async_load_network_settings(create_backup=True)
+        self._radio_mgr.chosen_backup = self._radio_mgr.backups[0]
+
+        self._radio_mgr.radio_type = new_radio_type
+        self._radio_mgr.device_path = new_device_settings[CONF_DEVICE_PATH]
+        self._radio_mgr.device_settings = new_device_settings
+        device_settings = self._radio_mgr.device_settings.copy()  # type: ignore[union-attr]
+
+        # Update the config entry settings
+        self._hass.config_entries.async_update_entry(
+            entry=self._config_entry,
+            data={
+                CONF_DEVICE: device_settings,
+                CONF_RADIO_TYPE: self._radio_mgr.radio_type.name,
+            },
+            options=self._config_entry.options,
+            title=name,
+        )
+        return False
+
+    async def async_restore_backup(self):
+        """Restore backup."""
+        assert self._radio_mgr.chosen_backup is not None
+
+        if self._radio_mgr.radio_type != RadioType.ezsp:
+            await self._radio_mgr.restore_backup(self._radio_mgr.chosen_backup)
+            return
+
+        # We have no way to partially load network settings if no network is formed
+        if self._radio_mgr.current_settings is None:
+            # Since we are going to be restoring the backup anyways, write it to the
+            # radio without overwriting the IEEE but don't take a backup with these
+            # temporary settings
+            temp_backup = _prevent_overwrite_ezsp_ieee(self._radio_mgr.chosen_backup)
+            await self._radio_mgr.restore_backup(temp_backup, create_new=False)
+            await self._radio_mgr.async_load_network_settings()
+
+            assert self._radio_mgr.current_settings is not None
+
+        if (
+            self._radio_mgr.current_settings.node_info.ieee
+            == self._radio_mgr.chosen_backup.node_info.ieee
+            or not self._radio_mgr.current_settings.network_info.metadata["ezsp"][
+                "can_write_custom_eui64"
+            ]
+        ):
+            # No point in prompting the user if the backup doesn't have a new IEEE
+            # address or if there is no way to overwrite the IEEE address a second time
+            await self._radio_mgr.restore_backup(self._radio_mgr.chosen_backup)
+
+            return
+
+        backup = self._radio_mgr.chosen_backup
+
+        backup = _allow_overwrite_ezsp_ieee(backup)
+
+        # If the user declined to overwrite the IEEE *and* we wrote the backup to
+        # their empty radio above, restoring it again would be redundant.
+        await self._radio_mgr.restore_backup(backup)
+
+        return
```

Comment:
I suspect this function can be simplified since we're always going to do a full restore, including overwriting the address etc.

Reply:
Refactored in commit d2842d0